llarp/router/rc_gossiper.cpp

```diff
@@ -0,0 +1,68 @@
+#include <router/rc_gossiper.hpp>
+#include <messages/dht_immediate.hpp>
+#include <dht/messages/gotrouter.hpp>
+
+namespace llarp
+{
+  RCGossiper::RCGossiper()
+      : I_RCGossiper(), m_Filter(RouterContact::UpdateInterval)
+  {
+  }
+
+  void
+  RCGossiper::Init(ILinkManager* l)
+  {
+    m_LinkManager = l;
+  }
+
+  void
+  RCGossiper::Decay(llarp_time_t now)
+  {
+    m_Filter.Decay(now);
+  }
+
+  bool
+  RCGossiper::GossipRC(const RouterContact& rc)
+  {
+    // only distribute public routers
+    if(not rc.IsPublicRouter())
+      return false;
+    if(m_LinkManager == nullptr)
+      return false;
+    // check for filter hit
+    if(not m_Filter.Insert(rc.pubkey))
+      return false;
+    bool sent = false;
+    // unwarrented GRCM
+    DHTImmediateMessage m;
+    m.msgs.emplace_back(
+        new dht::GotRouterMessage(dht::Key_t{}, 0, {rc}, false));
+
+    m_LinkManager->ForEachPeer([&](ILinkSession* s) {
+      // ensure connected session
+      if(not(s && s->IsEstablished()))
+        return;
+      // check if public router
+      const auto other_rc = s->GetRemoteRC();
+      if(not other_rc.IsPublicRouter())
+        return;
+      // dont send it to the owner
+      if(other_rc.pubkey == rc.pubkey)
+        return;
```

Comment:
Sending it back to the owner can actually be useful: it lets the owner know that another router has accepted and processed it.  In lokid gossiping we try to send every 5 minutes until we see it reflected back to us so that we know that at least one other node accepted it, and once that reflected record comes back we fall back to once-per-hour broadcasting.

Reply:
addresssed